serialization/serializer.go

```diff
@@ -2,217 +2,170 @@ package serialization
-	cfgInbox            *types.SyncMailbox[types.SerializerConfig, bool]
+	mut                 sync.Mutex
+	ser                 types.PrometheusMarshaller
 	maxItemsBeforeFlush int
 	flushFrequency      time.Duration
 	queue               types.FileStorage
 	lastFlush           time.Time
 	logger              log.Logger
-	self                actor.Actor
 	// Every 1 second we should check if we need to flush.
 	flushTestTimer *time.Ticker
-	series         []*types.TimeSeriesBinary
-	meta           []*types.TimeSeriesBinary
-	msgpBuffer     []byte
 	stats          func(stats types.SerializerStats)
+	fileFormat     types.FileFormat
+	needsStop      atomic.Bool
+	newestTS       int64
+	seriesCount    int
+	metadataCount  int
 }
 
-func NewSerializer(cfg types.SerializerConfig, q types.FileStorage, stats func(stats types.SerializerStats), l log.Logger) (types.Serializer, error) {
+func NewSerializer(cfg types.SerializerConfig, q types.FileStorage, stats func(stats types.SerializerStats), l log.Logger) (types.PrometheusSerializer, error) {
 	s := &serializer{
 		maxItemsBeforeFlush: int(cfg.MaxSignalsInBatch),
 		flushFrequency:      cfg.FlushFrequency,
 		queue:               q,
-		series:              make([]*types.TimeSeriesBinary, 0),
 		logger:              l,
-		inbox:               actor.NewMailbox[*types.TimeSeriesBinary](),
-		metaInbox:           actor.NewMailbox[*types.TimeSeriesBinary](),
-		cfgInbox:            types.NewSyncMailbox[types.SerializerConfig, bool](),
 		flushTestTimer:      time.NewTicker(1 * time.Second),
-		msgpBuffer:          make([]byte, 0),
 		lastFlush:           time.Now(),
 		stats:               stats,
+		fileFormat:          types.AlloyFileVersionV1,
+		ser:                 v1.GetSerializer(),
 	}
 
 	return s, nil
 }
-func (s *serializer) Start() {
-	// All the actors and mailboxes need to start.
-	s.self = actor.Combine(actor.New(s), s.inbox, s.metaInbox, s.cfgInbox).Build()
-	s.self.Start()
-}
 
-func (s *serializer) Stop() {
-	s.self.Stop()
-}
+// SendMetric adds a metric to the serializer. Note that we need to inject external labels here since once they are written to disk the prompb.TimeSeries bytes should be treated
+// as immutable.
+func (s *serializer) SendMetric(ctx context.Context, l labels.Labels, t int64, v float64, h *histogram.Histogram, fh *histogram.FloatHistogram, externalLabels map[string]string) error {
+	s.mut.Lock()
+	defer s.mut.Unlock()
 
-func (s *serializer) SendSeries(ctx context.Context, data *types.TimeSeriesBinary) error {
-	return s.inbox.Send(ctx, data)
-}
+	if t > s.newestTS {
+		s.newestTS = t
+	}
 
-func (s *serializer) SendMetadata(ctx context.Context, data *types.TimeSeriesBinary) error {
-	return s.metaInbox.Send(ctx, data)
-}
+	err := s.ser.AddPrometheusMetric(t, v, l, h, fh, externalLabels)
+	if err != nil {
+		return err
+	}
+	s.seriesCount++
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
```

Comment:
This comment mentions flush duration but the following code doesn't do anything with flush duration